Adding the ability to force 4DOF minimization in the point-to-plane.

kubelvla opened the pull request:
This adds the possibility to set the point-to-plane to minimize only x,y,z and yaw values.

ethzasl-jenkins commented:
Can one of the admins verify this patch?

simonpierredeschenes commented:
ok to test

YoshuaNava commented on pointmatcher/DataPointsFilters/Geometry.h:
```diff
@@ -0,0 +1,82 @@
+	virtual ~GeometryDataPointsFilter() {};
+	virtual DataPoints filter(const DataPoints& input);
+	virtual void inPlaceFilter(DataPoints& cloud);
+};
```
Given that the GeometryFilter files are introduced by a different PR, can you remove them from here?

kubelvla replied:
Yep, this is a little mess on my side, I was committing the filter code to this branch in my forked repo not realizing that this PR is still hanging here. There was a problem with the PointToGaussian minimizer which one of our students just fixed, so I hope this PR will get moving soon.

kubelvla replied:
And will try to remove the filter files without destroying my repo...

YoshuaNava commented on pointmatcher/ErrorMinimizers/PointToGaussian.cpp:
```diff
@@ -56,9 +56,29 @@ typedef Parametrizable::ParametersDoc ParametersDoc;
 	PointToPlaneErrorMinimizer<T>(PointToGaussianErrorMinimizer::availableParameters(), params),
-	confidenceInPenalties(Parametrizable::get<T>("confidenceInPenalties")),
-	force2D(Parametrizable::get<bool>("force2D"))
+	//confidenceInPenalties(Parametrizable::get<T>("confidenceInPenalties")),
```
Why is this line commented? Are these the penalties described in your Subarctic forests mapping paper?

kubelvla replied:
This will be fixed during the summer I hope. The point-to-gaussian minimizer was sleeping during our work on SubT, but it is being worked on right now. There was an issue with numerical stability causing the whole thing to diverge (somehow the computers still don't like dividing by zero, dunno what's the problem...)

YoshuaNava commented:
@kubelvla Is this feature still planned to reach master?

I imagine it being used in a SLAM context, to optimize XYZ+Yaw with ICP, and the rest with an IMU, which sounds quite useful!

Is this PR also related to what's shown in [this video](https://www.youtube.com/watch?v=t_cBdiPQ1R8) ?

YoshuaNava commented on pointmatcher/ErrorMinimizers/PointToGaussian.cpp:
```diff
@@ -56,9 +56,29 @@ typedef Parametrizable::ParametersDoc ParametersDoc;
+	{
+		LOG_INFO_STREAM("PointMatcher::PointToGaussianErrorMinimizer - minimization will be in full 6DOF.");
+	}
 }
```
The actual implementation of 4DOF opt. is missing from this file?

kubelvla replied:
Yep, it is quite straight-forward for point-to-plane, we need to figure out what it actually means for point-to-gaussian.

YoshuaNava commented on pointmatcher/ErrorMinimizers/PointToPlane.cpp:
```diff
@@ -154,7 +192,22 @@ typename PointMatcher<T>::TransformationParameters PointToPlaneErrorMinimizer<T>
 		// Compute cross product of cross = cross(reading X normalRef)
-		const Matrix cross = this->crossProduct(mPts.reading.features, normalRef);
+		Matrix cross;
+		Matrix Gamma(3,3);
```
Variable doesn't follow camelCasing style

kubelvla replied:
Will fix that.

YoshuaNava commented on pointmatcher/ErrorMinimizers/PointToPlane.cpp:
```diff
@@ -154,7 +192,22 @@ typename PointMatcher<T>::TransformationParameters PointToPlaneErrorMinimizer<T>
 		assert(normalRef.rows() > 0);
 
 		// Compute cross product of cross = cross(reading X normalRef)
```
This comment is repeated below, and it points at the math operation, but doesn't say what's the goal of it. Could it be rephrased in a way in which we know what's the goal, and then the next comments explain some of the steps if needed?

kubelvla replied:
Also goes to my TODO list. I'll put there reference to Francois' publication with equation numbers.

YoshuaNava commented on pointmatcher/ErrorMinimizers/PointToPlane.cpp:
```diff
@@ -154,7 +192,22 @@ typename PointMatcher<T>::TransformationParameters PointToPlaneErrorMinimizer<T>
+		   	//VK: Instead for "cross" as in 3D, we need only a dot product with the Gamma factor for 4DOF
+			Gamma << 0,-1, 0,
+			         1, 0, 0,
+			         0, 0, 0;
```
Where does this gamma factor come from? What's its physical interpretation?

If it's described in a paper, can we mention that somewhere?

kubelvla replied:
Well it is described in a bunch of wrinkled hand-written notes on my desk. I'll think how to put it somewhere.

YoshuaNava commented on pointmatcher/ErrorMinimizers/PointToPlane.cpp:
```diff
@@ -203,14 +256,29 @@ typename PointMatcher<T>::TransformationParameters PointToPlaneErrorMinimizer<T>
-				transform = Eigen::AngleAxis<T>(x.head(3).norm(),x.head(3).normalized());
+				if (!force4DOF)
+				{
+					transform = Eigen::AngleAxis<T>(x.head(3).norm(), x.head(3).normalized()); //x=[alpha,beta,gamma,x,y,z]
```
Related to the above comments, this block hints at Gamma = yaw.

Can this notation be defined before?

kubelvla replied:
The intention was that Gamma would be the matrix and gamma would be the error rotation vector z component (sort of delta yaw). I guess I'll call the previous matrixGamma and this one stay gamma.

YoshuaNava commented on pointmatcher/ErrorMinimizers/PointToPlane.h:
```diff
@@ -69,11 +69,14 @@ struct PointToPlaneErrorMinimizer: public PointMatcher<T>::ErrorMinimizer
         return {
-        	{"force2D", "If set to true(1), the minimization will be force to give a solution in 2D (i.e., on the XY-plane) even with 3D inputs.", "0", "0", "1", &P::Comp<bool>}
-            };
+                {"force2D", "If set to true(1), the minimization will be force to give a solution in 2D (i.e., on the XY-plane) even with 3D inputs.", "0", "0", "1", &P::Comp<bool>},
```
forced*

YoshuaNava commented:
I have a bit of a conceptual question: how possible is it to make the ICP optimization operate in arbitrary sub-spaces?

For example, let's say that we want to operate only in x,y,z or only yaw. Would it be very different to implement that?

kubelvla commented:
That depends on the error function you wish to minimize. I've only played with point-to-plane because it has a nice aproximative solution which is easy to modify. In the point-to-point, I presume that modifying the rotation part would be more challenging (but maybe, if we dropped the closed-form solution and replaced it with something aproximative it would become easier), but haven't tried that yet. Removing rotation altogether (doing only x,y,z) from the point-to-point and point-to-plane is straightforward. We are actually exploring all this this summer, mainly focusing on point-to-gaussian.

kubelvla commented:
@vdsmax Could you please eventually add your fix for the point-to-gaussian that you did with the additional measurement noise?

YoshuaNava commented:
@kubelvla Could this feature be integrated into master for the point to plan, and then bring in the implementation for the other error minimizers?

YoshuaNava commented:
@kubelvla Could this feature be integrated into master for the point to plan, and then bring in the implementation for the other error minimizers?

vdsmax commented:
> @vdsmax Could you please eventually add your fix for the point-to-gaussian that you did with the additional measurement noise?

@kubelvla i will add the change today

kubelvla commented:
@YoshuaNava Yes, that is possible to be done.

YoshuaNava commented on pointmatcher/ErrorMinimizers/PointToGaussian.cpp:
```diff
@@ -117,21 +118,23 @@ typename PointToGaussianErrorMinimizer<T>::ErrorElements PointToGaussianErrorMin
-		// Convert a vector into a matrix
-		const Matrix matEigVectors = Eigen::Map<const Matrix>(eigVectors.col(i).data(), dim, dim);
-		normals.block(0, dim * i, dim, dim) = matEigVectors;
+    for (long i = 0; i < mPts_const.reference.features.cols(); ++i) {
```
@vdsmax Can you add some high level comments describing what is being done here?

pomerlef replied:
@vdsmax ping

YoshuaNava commented on pointmatcher/ErrorMinimizers/PointToGaussian.cpp:
```diff
@@ -117,21 +118,23 @@ typename PointToGaussianErrorMinimizer<T>::ErrorElements PointToGaussianErrorMin
+
+        normals.block(0, dim * i, dim, dim) = matEigVectors.transpose();
+
+        // A eigen value of zero will have infinite weight, that's why we add a noise constant for the weight
```
```suggestion
        // An eigen value of zero will have infinite weight in the cost function, that's why we add a noise constant for the weight
```

YoshuaNava commented on pointmatcher/ErrorMinimizers/PointToGaussian.cpp:
```diff
@@ -117,21 +118,23 @@ typename PointToGaussianErrorMinimizer<T>::ErrorElements PointToGaussianErrorMin
-		assert((eigValues.col(i).array() != 0.0).any());
-		mPts.weights.block(0, dim * i, 1, dim) = eigValues.col(i).array().inverse().transpose();
-	}
+        // Convert a vector into a matrix
```
Why is this done?

YoshuaNava commented on pointmatcher/ErrorMinimizers/PointToGaussian.cpp:
```diff
@@ -117,21 +118,23 @@ typename PointToGaussianErrorMinimizer<T>::ErrorElements PointToGaussianErrorMin
+        // Convert a vector into a matrix
+        const Matrix matEigVectors = Eigen::Map<const Matrix>(eigVectors.col(i).data(), dim, dim);
+
+        normals.block(0, dim * i, dim, dim) = matEigVectors.transpose();
```
Are you overwriting the normals?

YoshuaNava commented on pointmatcher/ErrorMinimizers/PointToGaussian.cpp:
```diff
@@ -117,21 +118,23 @@ typename PointToGaussianErrorMinimizer<T>::ErrorElements PointToGaussianErrorMin
+
+        // A eigen value of zero will have infinite weight, that's why we add a noise constant for the weight
+        // Same solution as Bosse did with his Zebedee platefrom
+        mPts.weights.block(0, dim * i, 1, dim) = (pow(eigValues.col(i).array()+noiseSensor*noiseSensor,1/2)).inverse().transpose();
```
Pow might not be the fastest way to compute a square in this context: https://baptiste-wicht.com/posts/2017/09/cpp11-performance-tip-when-to-use-std-pow.html

YoshuaNava commented on pointmatcher/ErrorMinimizers/PointToGaussian.h:
```diff
@@ -84,6 +85,7 @@ struct PointToGaussianErrorMinimizer: public PointToPlaneErrorMinimizer<T>
     //const T confidenceInPenalties;
     const bool force2D;
     const bool force4DOF;
+    const float noiseSensor;
```
I hope I'm not being too picky. If that's the case feel free to disregard this comment

noiseSensor sounds like a noise sensor, a device that detects noise.

sensorNoise would indicate the expected noise of a sensor. I think it is a more accurate term.

Now, what is this quantity? The variance, standard deviation of the noise?

YoshuaNava commented on pointmatcher/ErrorMinimizers/PointToGaussian.cpp:
```diff
@@ -117,21 +118,23 @@ typename PointToGaussianErrorMinimizer<T>::ErrorElements PointToGaussianErrorMin
-		normals.block(0, dim * i, dim, dim) = matEigVectors;
+    for (long i = 0; i < mPts_const.reference.features.cols(); ++i) {
+        mPts.reading.features.block(0, dim * i, dim + 1, dim) = mPts_const.reading.features.col(i).replicate(1, dim);
+        mPts.reference.features.block(0, dim * i, dim + 1, dim) = mPts_const.reference.features.col(i).replicate(1, dim);
```
Why are you replicating sections of these matrices? Would a View or const View suffice?

YoshuaNava commented:
@kubelvla Can you update this branch with the latest master?

kubelvla commented:
@YoshuaNava Merged master into this branch. Yet, work on the point-to-gaussian is in progress, we will be working on that this summer. The minimizer needs to be tested more. Should we keep this PR hanging to keep this conversation, or should I close it for now?

YoshuaNava commented:
> @YoshuaNava Merged master into this branch. Yet, work on the point-to-gaussian is in progress, we will be working on that this summer. The minimizer needs to be tested more. Should we keep this PR hanging to keep this conversation, or should I close it for now?

I think we should keep it.

pomerlef commented:
@kubelvla is that PR ready for a merge?

pomerlef commented:
@kubelvla there seems to be some effort in that PR. Can we do something to progress?

kubelvla commented:
@pomerlef How's Point-to-Gaussian behaving? @vdsmax @simonpierredeschenes have you been using that after I left?

simonpierredeschenes commented:
I did not use it on my side unfortunately, but I think @vdsmax did

kubelvla commented:
@pomerlef The 4DOF part is just an implicit part of this pull request, since it is a part of the point-to-plane minimizer. I am not sure if there actually was any more work done on the point-to-gaussian minimization after Max had played with that for some time (was it the Blimp project he used that for?) So the question is whether we want this to be released (with 4DOF or without), or if you want to put some student on it (and then the question is how to generate the gaussian "points" - it was from GPS at some point, that was the motivation for Bab, right?)

pomerlef commented:
I'm actually not sure what to do with this PR. I'll close it. Just re-open it if someone wants to work on it (but from the conversation I doubt it).
